xc7/utils/prjxray_create_edges.py

```diff
@@ -285,11 +285,273 @@ def __init__(self, pins=None, tracks=None):
                 connection_database.get_track_model, which builds the Tracks
                 models and the graph node list.
         """
+        self.conn = conn
         self.pins = pins
         self.tracks = tracks
         assert (self.pins is not None) ^ (self.tracks is not None)
 
-    def connect_at(self, loc, other_connector):
+    def get_pip_switch(
+            self, src_wire_pkey, pip_pkey, dest_wire_pkey, switch_pkey
+    ):
+        """ Return the switch_pkey for the given connection.
+
+        Selects either normal or backward switch from pip, or if switch is
+        already known, returns known switch.
+
+        It is not valid to provide a switch and provide src/dest/pip arguments.
+
+        Arguments
+        ---------
+        src_wire_pkey : int
+            Source wire row primary key.  May be None if switch_pkey is not
+            None.
+        pip_pkey : int
+            Pip connecting source to destination wire.  May be None if
+            switch_pkey is not None.
+        dest_wire_pkey : int
+            Destination wire row primary key.  May be None if switch_pkey
+            is not None.
+        switch_pkey : Int
+            Switch row primary key, can be used if switch_pkey is already
+            known (e.g. synthetic edge).  If switch_pkey is not None, other
+            arguments should be None to avoid ambiguity.
+
+        Returns
+        -------
+        Switch row primary key to connect through specified pip.
+
+        """
+        if switch_pkey is not None:
```

Comment:
In the case of synthetic edges (e.g. constant network, or bypasses), a lookup cannot succeed.  In these cases a switch_pkey is provided, rather than attempting a lookup that will fail.